internal/validator/validator.go

```diff
@@ -0,0 +1,317 @@
+			claim, _, err := merkle.CreateProofs(nil, MAX_OUTPUT_TREE_HEIGHT)
+			if err != nil {
+				return nil, nil, fmt.Errorf(
+					"failed to create proofs for epoch %v of application %v. %w",
+					epoch.Index, epoch.AppAddress, err,
+				)
+			}
+			return &claim, nil, nil
+		}
+	} else {
+		// if epoch has outputs, calculate a new claim and proofs
+		var previousOutputs []Output
+		if previousEpoch != nil {
+			// get all outputs created before the current epoch
+			previousOutputs, err = v.repository.GetOutputsProducedInBlockRange(
+				ctx,
+				epoch.AppAddress,
+				v.inputBoxDeploymentBlock,
+				previousEpoch.LastBlock,
+			)
+			if err != nil {
+				return nil, nil, fmt.Errorf(
+					"failed to get all outputs of application %v before epoch %d. %w",
+					epoch.AppAddress, epoch.Index, err,
+				)
+			}
+		}
+		// the leaves of the Merkle tree are the Keccak256 hashes of all the
+		// outputs
+		leaves := make([]Hash, 0, len(epochOutputs)+len(previousOutputs))
+		for idx := range previousOutputs {
+			if previousOutputs[idx].Hash == nil {
+				// should never happen
+				return nil, nil, fmt.Errorf(
+					"missing hash of output %d of epoch %d",
+					previousOutputs[idx].Index, epoch.Index,
+				)
+			}
+			leaves = append(leaves, *previousOutputs[idx].Hash)
+		}
+		for idx := range epochOutputs {
+			hash := crypto.Keccak256Hash(epochOutputs[idx].RawData[:])
+			// update outputs with their hash
+			epochOutputs[idx].Hash = &hash
+			// add them to the leaves slice
+			leaves = append(leaves, hash)
+		}
+
+		claim, proofs, err := merkle.CreateProofs(leaves, MAX_OUTPUT_TREE_HEIGHT)
+		if err != nil {
+			return nil, nil, fmt.Errorf(
+				"failed to create proofs for epoch %d of application %v. %w",
+				epoch.Index, epoch.AppAddress, err,
+			)
+		}
+
+		// update outputs with their proof
+		for idx := range epochOutputs {
+			start := epochOutputs[idx].Index * MAX_OUTPUT_TREE_HEIGHT
+			end := (epochOutputs[idx].Index * MAX_OUTPUT_TREE_HEIGHT) + MAX_OUTPUT_TREE_HEIGHT
+			epochOutputs[idx].OutputHashesSiblings = proofs[start:end]
+		}
+		return &claim, epochOutputs, nil
+	}
+	// if there are no outputs and there is a previous epoch, return its claim
+	return previousEpoch.ClaimHash, nil, nil
+}
+
+// ValidatorService extends the Validator utility by executing it with a polling
+// strategy. It implements the `services.Service` interface.
+type ValidatorService struct {
+	validator       *Validator
+	pollingInterval time.Duration
+}
+
+func NewValidatorService(
+	repo ValidatorRepository,
+	inputBoxDeploymentBlock uint64,
+	pollingInterval time.Duration,
+) *ValidatorService {
+	service := &ValidatorService{pollingInterval: pollingInterval}
+	service.validator = NewValidator(repo, inputBoxDeploymentBlock)
+	return service
+}
+
+func (s *ValidatorService) String() string {
+	return "validator"
+}
+
+func (s *ValidatorService) Start(ctx context.Context, ready chan<- struct{}) error {
```

Comment:
We can start using the Poller. It is already merged.
Add a TODO somewhere if you don't do it in this PR.

Reply:
implementing Poller now would require small adjustments in several files, so I will refrain from doing it here to keep this PR leaner. But this is already planned for post-alpha release, don't worry 👍🏽